lib/motors/controller.cpp

```diff
@@ -2,8 +2,128 @@
+      break;
+    case 0xFF10:
+      logger_.log(core::LogLevel::kFatal,
+                  "ERROR_CONTROLLER_OVERTEMPERATURE: Controller overtemperature");
+      break;
+    case 0x3210:
+      logger_.log(core::LogLevel::kFatal,
+                  "ERROR_DC_LINK_OVERVOLTAGE: Power supply voltage too high");
+      break;
+    default:
+      logger_.log(core::LogLevel::kFatal,
+                  "GENERIC_ERROR: Unspecific error occurred with code %i",
+                  error_code);
+      break;
+  }
+}
+
+controllerStatus Controller::processWarningMessage(const std::uint8_t warning_code)
+{
+  controllerStatus priority_error = controllerStatus::kNominal;
+
+  // Flag specific warnings in binary
+  if ((warning_code & 0b0000'0001) == 0b000'0000'0001) {
+    logger_.log(core::LogLevel::kInfo, "Controller Temperature Exceeded");
+    priority_error = controllerStatus::kControllerTemperatureExceeded;
+  }
+  if ((warning_code & 0b0000'0000'0010) == 0b0000'0000'0010) {
+    logger_.log(core::LogLevel::kFatal, "Motor Temperature Exceeded");
+    priority_error = controllerStatus::kUnrecoverableWarning;
+  }
+  if ((warning_code & 0b0000'0000'0100) == 0b0000'0000'0100) {
+    logger_.log(core::LogLevel::kFatal, "DC link under voltage");
+    priority_error = controllerStatus::kUnrecoverableWarning;
+  }
+  if ((warning_code & 0b0000'0000'1000) == 0b0000'0000'1000) {
+    logger_.log(core::LogLevel::kFatal, "DC link over voltage");
+    priority_error = controllerStatus::kUnrecoverableWarning;
+  }
+  if ((warning_code & 0b0000'0001'0000) == 0b0000'0001'0000) {
+    logger_.log(core::LogLevel::kFatal, "DC over priority_errorent");
+    priority_error = controllerStatus::kUnrecoverableWarning;
+  }
+  if ((warning_code & 0b0000'0010'0000) == 0b0000'0010'0000) {
+    logger_.log(core::LogLevel::kFatal, "Stall protection active");
+    priority_error = controllerStatus::kUnrecoverableWarning;
+  }
+  if ((warning_code & 0b0000'0100'0000) == 0b0000'0100'0000) {
+    logger_.log(core::LogLevel::kFatal, "Max velocity exceeded");
+    priority_error = controllerStatus::kUnrecoverableWarning;
+  }
+  if ((warning_code & 0b0000'1000'0000) == 0b0000'1000'0000) {
+    logger_.log(core::LogLevel::kFatal, "BMS Proposed Power");
+    priority_error = controllerStatus::kUnrecoverableWarning;
+  }
+  if ((warning_code & 0b0001'0000'0000) == 0b0001'0000'0000) {
+    logger_.log(core::LogLevel::kFatal, "Capacitor temporature exceeded");
+    priority_error = controllerStatus::kUnrecoverableWarning;
+  }
+  if ((warning_code & 0b0010'0000'0000) == 0b0010'0000'0000) {
+    logger_.log(core::LogLevel::kFatal, "I2T protection");
+    priority_error = controllerStatus::kUnrecoverableWarning;
+  }
+  if ((warning_code & 0b0100'0000'0000) == 0b0100'0000'0000) {
+    logger_.log(core::LogLevel::kFatal, "Field weakening active");
+    priority_error = controllerStatus::kUnrecoverableWarning;
+  }
+
+  return priority_error;
+}
```

Comment:
So this function needs some work...

1. I'm not a massive fan of the mutable `priority_error` here but I see why we would do it like this (just for logging); how about this: catch the errors in decreasing order of priority (reverse from what you're doing here), log the most important error (i.e. in english) AND the full error code (this will allow us to debug other errors/warnings that may have occurred). 
2. Don't use binary; hex does everything you need.
3. `if ((warning_code & MASK) == MASK) { /* ... */ }` is equivalent to `if (warning_code & MASK) { /* ... */  }`

Pytting these together, here's what I think this should look like:

```cpp
if (warning_code & 0x40) {
  logger_.log(core::LogLevel::kFatal, "Controller warning: Field weakening active (code: %x)", warning_code);
  return ControllerStatus::kUnrecoverableWarning;
}
if (warning_code & 0x20) {
  // ...
}
// ...
return ControllerStatus::kNominal;
```

Reply:
I agree this is cleaner, will get this implemented.

Reply:
did not consider this, perhaps could roll back processWarningMessage to last commit and just change the bitwise conditions to be neater. I suppose a more readable output is probable preferable.

Reply:
> My problem with this is that we then have to look it up in the manual in the case of 2 warnings. We should log all warnings, especially fatal ones

Hm, you're kind of right? There is a problem here, though: Logging is technically an I/O operation so I don't really want all of this work to happen here. Now, this is a problem for telemetry but I'm a bit confused about what they've been up to this year because there literally isn't anything in this repository yet. @kshxtij ?

Reply:
They're working on the UI end right now because they don't need pod-side code beyond stuff that plugs directly into ros.

Reply:
Cool, Tristan just rolled back the function to the mutable variable version, think this is the best outcome

Reply:
Why does the UI not live in this repo? 

Also, this is exactly what scares me! We have a failure and what do we do? We decide to send stupid things over the network.